### PR TITLE
CEDS-5830 draft

### PR DESCRIPTION
**Pull request by magicaljellybeans**
- Addition of AmendmentHelper
- Addition of amendKeys to PointerRecord class (partially complete)

TODO add logic that fetches all leaf (only) pointer records when a parent level pointer comes through (via the DiffTool)
Capture any logic from valueAdded and other methods on case classes extending Amendment/AmendmentOp traits

### CODE REVIEW

File: app/controllers/helpers/AmendmentHelper.scala [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+      val diff = amendedDeclaration.createDiff(originalDeclaration)
+      diff.flatMap { alteredField =>
+        PointerPatterns.expandPointer(Pointer.apply(alteredField.fieldPointer), originalDeclaration, amendedDeclaration)
+      }
+    }
+
+    val pointersAndRecords = amendedPointers.map(pointer => (pointer, PointerRecord.pointersToPointerRecords(pointer.pattern)))
+
+    pointersAndRecords.flatMap { case (pointer, record) =>
+      record.amendKey.fold {
```
Comment (udinobi) [inline app/controllers/helpers/AmendmentHelper.scala:44]: The `amendedPointers.map(` loop seems superfluous. Cannot L41-L44 be reduced to...
```
amendedPointers.flatMap { pointer =>
  val record = PointerRecord.pointersToPointerRecords(pointer.pattern)
  record.amendKey.fold {
```

File: app/forms/common/YesNoAnswer.scala [hunk 2]
```diff
@@ -35,28 +33,15 @@ case class YesNoAnswer(answer: String) extends Ordered[YesNoAnswer] with Amendme
...
 object YesNoAnswer {
 
   implicit val format: OFormat[YesNoAnswer] = Json.format[YesNoAnswer]
 
   private lazy val parties = s"${ExportsDeclaration.pointer}.${Parties.pointer}"
-  private lazy val transport = s"${ExportsDeclaration.pointer}.${Transport.pointer}"
 
-  lazy val mappingsForAmendment = Map(
-    s"$parties.personPresentingGoodsDetails.eori" -> "declaration.summary.parties.eidr",
-    s"$transport.expressConsignment" -> "declaration.summary.transport.expressConsignment"
```
Comment (udinobi) [inline app/forms/common/YesNoAnswer.scala:58]: Just to be sure... don't we need any more the `expressConsignment` mapping?

File: app/views/helpers/PointerPatterns.scala [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+  val expandAdditionalActors = (p: Pointer, orig: ExportsDeclaration, amend: ExportsDeclaration) =>
+    Seq(Pointer(p.sections :+ PointerSection("eori", FIELD)), Pointer(p.sections :+ PointerSection("type", FIELD)))
+
+  val expandCusCode = (p: Pointer, orig: ExportsDeclaration, amend: ExportsDeclaration) =>
+    Seq(Pointer(p.sections :+ PointerSection("cusCode", FIELD)))
+
+  val expandConsignorDetails = (p: Pointer, orig: ExportsDeclaration, amend: ExportsDeclaration) =>
+    Seq(Pointer(p.sections :+ PointerSection("eori", FIELD)))
+
+  val expandDetails = (p: Pointer, orig: ExportsDeclaration, amend: ExportsDeclaration) => Seq(Pointer(p.sections :+ PointerSection("eori", FIELD)))
```
Comment (udinobi) [inline app/views/helpers/PointerPatterns.scala:37]: What's this for? I do not see differences with `expandConsignorDetails`

Comment (glidester) [reply app/views/helpers/PointerPatterns.scala:37]: Yes good spot, looks like we can just reuse the generic `expandDetails` function.

File: app/controllers/helpers/AmendmentHelper.scala (hunk 1 above)
Comment (glidester) [reply app/controllers/helpers/AmendmentHelper.scala:44]: Yes, will rationalise it down to a single loop :+1:

File: app/forms/common/YesNoAnswer.scala (hunk 2 above)
Comment (glidester) [reply app/forms/common/YesNoAnswer.scala:58]: We don't need this construct at all now. Will remove it from the code completely :+1:

**Review by udinobi — APPROVED** [review]

**Pull request merged**